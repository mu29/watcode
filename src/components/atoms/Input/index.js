import React from 'react'
import { PropTypes } from 'prop-types'
import styled, { css } from 'styled-components'
import { width, border, color, primaryColor, fontFamily, fontSize, palette } from 'services/style'

const errorColor = ({ error, theme }) => error && `1px solid ${palette('red.default')({ theme })}`

const styles = css`
  width: ${width};
  color: ${color};
  font-family: ${fontFamily};
  font-size: ${fontSize};
  border: ${border};
  border-radius: 0.0625rem;
  background: transparent;
  padding: 0.75rem;
  transition: border 0.3s;

  &:hover {
    outline: none;
  }

  &:active, &:focus {
    outline: none;
    border: 0.0625rem solid ${primaryColor};
  }

  &::placeholder {
    filter: opacity(50%);
  }

  &:-webkit-autofill,
  &:-webkit-autofill:hover,
  &:-webkit-autofill:focus,
  &:-webkit-autofill:active {
    -webkit-transition: "color 9999s ease-out, background-color 9999s ease-out";
    -webkit-transition-delay: 9999s;
  }
`

const StyledInput = styled.input`
  ${styles}
  border: ${errorColor};

  &:active, &:focus {
  border: ${errorColor};
  }
`
const StyledTextArea = styled.textarea`${styles}`

const Input = (props) => {
  switch (props.type) {
    case 'textarea':
      return <StyledTextArea { ...props } />
    default:
      return <StyledInput { ...props } />
  }
}

Input.propTypes = {
  type: PropTypes.string,
  width: PropTypes.number,
  border: PropTypes.string,
  block: PropTypes.bool,
  color: PropTypes.string,
  primaryColor: PropTypes.string,
  font: PropTypes.string,
  fontSize: PropTypes.number,
}

Input.defaultProps = {
  type: 'text',
  width: undefined,
  border: 'gray.40',
  block: false,
  font: 'default',
  fontSize: 16,
  color: 'gray.90',
  primaryColor: 'yellow.default',
}

export default Input
