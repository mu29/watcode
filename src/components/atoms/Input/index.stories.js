import React from 'react'
import { storiesOf } from '@storybook/react'
import Input from '.'

storiesOf('Atoms/Input', module)
  .add('기본', () => (
    <Input />
  ))
  .add('도움 텍스트', () => (
    <Input placeholder="메일 주소를 입력해주세요" />
  ))
