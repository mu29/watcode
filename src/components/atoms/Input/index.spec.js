import React from 'react'
import Input from '.'

const component = (props = {}) => withTheme(<Input { ...props } />)
const wrap = (props = {}) => shallow(component(props)).dive()

describe('<Input />', () => {
  it('정상적으로 렌더링된다', () => {
    const wrapper = render(component())
    expect(wrapper).toMatchSnapshot()
  })

  it('임의의 속성이 들어온 경우 그려준다', () => {
    const wrapper = wrap({ name: 'foo' })
    expect(wrapper.props().name).toEqual('foo')
  })

  it('오류 메시지를 표시한다', () => {
    const wrapper = wrap({ error: 'foo error' })
    expect(wrapper).toMatchSnapshot()
  })

  it('기본으로 input 태그를 그려준다', () => {
    const wrapper = wrap()
    expect(wrapper.find('input')).toHaveLength(1)
  })

  it('type 속성이 textarea인 경우 textarea 태그를 그려준다', () => {
    const wrapper = wrap({ type: 'textarea' })
    expect(wrapper.find('textarea')).toHaveLength(1)
  })
})
