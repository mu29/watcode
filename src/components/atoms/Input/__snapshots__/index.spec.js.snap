// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> 오류 메시지를 표시한다 1`] = `
.c0 {
  color: #424242;
  font-family: 'Spoqa Han Sans','Sans-serif','Helvetica Neue','Nanum Gothic';
  font-size: 1rem;
  border: 0.0625rem solid #E0E0E0;
  border-radius: 0.0625rem;
  background: transparent;
  padding: 0.75rem;
  -webkit-transition: border 0.3s;
  transition: border 0.3s;
  border: 1px solid #FF8787;
}

.c0:hover {
  outline: none;
}

.c0:active,
.c0:focus {
  outline: none;
  border: 0.0625rem solid #F9D14C;
}

.c0::-webkit-input-placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0::-moz-placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0:-ms-input-placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0::placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0:-webkit-autofill,
.c0:-webkit-autofill:hover,
.c0:-webkit-autofill:focus,
.c0:-webkit-autofill:active {
  -webkit-transition: "color 9999s ease-out, background-color 9999s ease-out";
  -webkit-transition-delay: 9999s;
}

.c0:active,
.c0:focus {
  border: 1px solid #FF8787;
}

<input
  className="c0"
  color="gray.90"
  fontSize={16}
  type="text"
/>
`;

exports[`<Input /> 정상적으로 렌더링된다 1`] = `
.c0 {
  color: #424242;
  font-family: 'Spoqa Han Sans','Sans-serif','Helvetica Neue','Nanum Gothic';
  font-size: 1rem;
  border: 0.0625rem solid #E0E0E0;
  border-radius: 0.0625rem;
  background: transparent;
  padding: 0.75rem;
  -webkit-transition: border 0.3s;
  transition: border 0.3s;
}

.c0:hover {
  outline: none;
}

.c0:active,
.c0:focus {
  outline: none;
  border: 0.0625rem solid #F9D14C;
}

.c0::-webkit-input-placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0::-moz-placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0:-ms-input-placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0::placeholder {
  -webkit-filter: opacity(50%);
  filter: opacity(50%);
}

.c0:-webkit-autofill,
.c0:-webkit-autofill:hover,
.c0:-webkit-autofill:focus,
.c0:-webkit-autofill:active {
  -webkit-transition: "color 9999s ease-out, background-color 9999s ease-out";
  -webkit-transition-delay: 9999s;
}

<input
  className="c0"
  color="gray.90"
  fontSize={16}
  type="text"
/>
`;
